File: src/main/java/com/pandus/leetcode/solutions/google/BestTimeToBuyAndSellStock.java
package com.pandus.leetcode.solutions.google;

public class BestTimeToBuyAndSellStock {
    public int maxProfit(int[] prices) {
        int ans = 0;
        int buyPrice = prices[0];
        for (int i = 1; i < prices.length; i++) {
            if (prices[i] - buyPrice < ans) {
                buyPrice = Math.min(prices[i], buyPrice);
            }
            ans = Math.max(ans, prices[i] - buyPrice);
        }
        return ans;
    }
}

File: src/main/kotlin/com/pandus/leetcode/solutions/top/interview/BestTimeToBuyAndSellStock.kt
package com.pandus.leetcode.solutions.top.interview

class BestTimeToBuyAndSellStock {

    //Brute-force solution
    fun maxProfit1(prices: IntArray): Int {
        var maxProfit = 0
        val n = prices.size
        for (current in 0 until n) {
            for (futureDay in current + 1 until n) {
                val curPrice = prices[current]
                val nextPrice = prices[futureDay]
                if (curPrice < nextPrice) {
                    val tempProfit = nextPrice - curPrice
                    if (tempProfit > maxProfit) {
                        maxProfit = tempProfit
                    }
                }
            }
        }
        return maxProfit
    }


    fun maxProfit(prices: IntArray): Int {
        var maxProfit = 0
        var minPrice = Int.MAX_VALUE

        for (price in prices) {
            minPrice = minOf(minPrice, price)
            maxProfit = maxOf(maxProfit, price - minPrice)
        }
        return maxProfit
    }
}
